### PR TITLE
sl_thd WOKEN race

### PR DESCRIPTION
**Pull request by maloneya**
### Summary of this Pull Request (PR)

**Add description here.**
Intended to solve a race that was occurring that would cause threads to deadlock in the case where a thread is woken before it actually calls block.

though this solves my issue i may have missed some necessary changes elsewhere in parts of the scheduler my code isn't using. 

updated state sl_thd state machine:
    - runnable -> blocked: when the thread blocks on an operation
    - runnable -> woke: when a thread is runnable and is `sl_wakeup`ed by another thread
    - woke -> runnable: when a thread blocks `sl_block`s when it has been previously `sl_wakeup`ed

### Intent for your PR

Choose one (Mandatory):

- [ ] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [x ] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):
(Specify @<github.com username(s)> of the reviewers. Ex: @user1, @user2)

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [ x] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG
- [x ] Naming adhere's to the SG
- [x ] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [ x] I've made an attempt to remove all redundant code
- [x ] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x ] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [ x] I've commented appropriately where code is tricky
- [ x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):
voter/demo_app.rs
-

### CODE REVIEW

File: src/components/include/sl.h [hunk 1]
```diff
@@ -462,15 +462,16 @@ sl_cs_exit_schedule_nospin_arg(struct sl_thd *to)
 		}
 	}
 
-	assert(t->state == SL_THD_RUNNABLE);
+	//WOKE is subset of RUNNABLE so should be safe to dispatch to
```
Comment (gparmer) [inline src/components/include/sl.h]: `/* */` is the prescribed comment style for single-line comments.

Comment (gparmer) [reply src/components/include/sl.h]: I don't think that describing the WOKE state as a *subset* is correct.  Comment *could* read:

`/* WOKE threads are still runnable */`

Likely the extensible way to handle this is to have a function `sl_is_runnable(t)` that can encode which states include threads that can be dispatched to.

File: src/components/lib/sl/sl.c [hunk 2]
```diff
@@ -157,17 +157,24 @@ sl_thd_block_no_cs(struct sl_thd *t, sl_thd_state_t block_type, cycles_t timeout
 void
 sl_thd_block(thdid_t tid)
 {
+
```
Comment (gparmer) [inline src/components/lib/sl/sl.c]: Random empty line?

File: src/components/lib/sl/sl.c [hunk 3]
```diff
@@ -157,17 +157,24 @@ sl_thd_block_no_cs(struct sl_thd *t, sl_thd_state_t block_type, cycles_t timeout
...
 	struct sl_thd *t;
 
 	/* TODO: dependencies not yet supported */
 	assert(!tid);
 
 	sl_cs_enter();
 	t = sl_thd_curr();
-	if (sl_thd_block_no_cs(t, SL_THD_BLOCKED, 0)) {
+
```
Comment (gparmer) [inline src/components/lib/sl/sl.c]: Not really a new "paragraph", so no blank line?

File: src/components/lib/sl/sl.c [hunk 4]
```diff
@@ -157,17 +157,24 @@ sl_thd_block_no_cs(struct sl_thd *t, sl_thd_state_t block_type, cycles_t timeout
...
+	if (unlikely(t->state == SL_THD_WOKEN)) {
+		t->state = SL_THD_RUNNABLE;
+	}
+
```
Comment (gparmer) [inline src/components/lib/sl/sl.c]: absolutely no paragraph between different conditions.  No empty line.

**Comment by maloneya** [comment]
Style changes made. Went ahead ad added sl_is_runnable(t) to sl.h

**Comment by gparmer** [comment]
Sorry, but this isn't mergeable as is.  Do a `pull` as there are conflicts with the current tree in `src/components/lib/sl/sl.c`.

**Comment by phanikishoreg** [comment]
Sorry, this is not a complete fix for the reasons I've mentioned on slack: 

*******************
For AEP threads, we get multiple consecutive `blocked/unblocked` events from kernel and are very common (not edge cases). 

Ex: Lets say that a thread is blocked on `cos_rcv`, and an interrupt woke it up and it runs to completion (because it is higher prio than the currently executing thread). When the thread calls `cos_rcv` again, it activates its scheduler, which sees that the event is a `blocked` event (because that thread is blocked on cos_rcv at this point). At this point, scheduler sees that it's a duplicate blocked event (as in the thread is already blocked). If this repeats, that is multiple blocked events. 

Another case is multiple wake up events.. Ex: A thread is woken up by an interrupt, but hasn't finished executing (was preempted by another higher prio thread or because the thread ran out of budget), when the scheduler activates, it sees that this was a wakeup event. Imagine a case where this repeats. There will be multiple wakeup events. I'm trying to think if there were any additional edge cases, I'll post them if I remember of any. 

@maloneya’s PR  brings back the old code for WOKEN state (without special handling for scheduling events from kernel, seem to work for his setup and I assume because there are no kernel/scheduler events). But that wouldn't work well in a complex system. We need special handling for scheduler events from kernel, and have WOKEN state for user-level races as you suggested before.
*******************

I've made changes to allow WOKEN state for user-level races and have separate block/wakeup API for redundant scheduling events. 
I will be able to submit a PR after @maloneya helps me test and confirm that the deadlock doesn't occur. I've tested the sl library/sched-component unit-tests we have in the mainline and it seems to work. (We need a unit test framework for lock API.)

If it's okay, please close this PR because merging this will not work well for many use-cases we have.

**Comment by gparmer** [comment]
OK, @phanikishoreg work with @maloneya to put together a PR that does fix the requisite issues.  Regardless, we need locks working, so if this is going to take a while, let me know.

**Pull request closed without merging**